casper/src/main/resources/PoS.rhox

```diff
@@ -63,626 +66,646 @@ in {
+                      resultCh!(state, (false, "Bond is greater than maximum!"))
+                    } else {
+                      deposit!(deployerId, amount, state.get("activeValidators"), *depositCh) |
+                      for (@depositResult <- depositCh) {
+                        match depositResult {
+                          (true, _) => {
+                            resultCh!(state.set("allBonds",state.get("allBonds").set(userPk,amount)),
+                                      depositResult)
+                          }
+                          (false, errorMsg) => {
+                            resultCh!(state, (false, "Bond deposit failed: " ++ errorMsg))
                           }
                         }
                       }
-                    } else {
-                      return!((false, "Invalid system auth token"))
                     }
                   }
                 }
               } |
-
-              contract transferFromBlockSender(@deployerRevAddress, @refundAmount, returnCh) = {
-                new getBlockData(`rho:block:data`),
-                    blockDataCh,
-                    indexSender,
-                    unfAuthKeyCh,
-                    tmpCh in {
+              // Public method allowing validators to withdraw their bonded and accumulated funds.
+              contract PoS(@"withdraw", @deployerId, returnCh) = {
+                new userCh, processCh, blockDataCh,
+                    getBlockData(`rho:block:data`)
+                in {
+                  // Consumes state and updates withdrawers map.
+                  runMVar!(*stateCh, *processCh, *returnCh) |
                   getBlockData!(*blockDataCh) |
-                  for (_, _, @sender <- blockDataCh) {
-                    for (@state <<- stateCh) {
-                      @ListOps!("indexOf", state.get("activeValidators"), sender, *indexSender) |
-                      for (@idx <- indexSender) {
-                        for (@vaults <<- perActiveValidatorPosVaultsCh) {
-                          if (idx == -1) {
-                            returnCh!((false, "Sender is not an active validator"))
-                          } else {
-                            match vaults.nth(idx) {
-                              (unf, _, activeValidatorVault) => {
-                                @RevVault!("unforgeableAuthKey", unf, *unfAuthKeyCh) |
-                                for (@authKey <- unfAuthKeyCh) {
-                                  @activeValidatorVault!("transfer", deployerRevAddress, refundAmount, authKey, *returnCh)
+                  getUser!(deployerId, *userCh) |
+                  for (@state, resultCh   <- processCh;
+                       @userPk            <- userCh;
+                       @blockNumber, _, _ <- blockDataCh) {
+                    if (state.get("allBonds").contains(userPk)) {
+                      // The withdwal goes into effect at the sart of the next epoch.
```

Comment:
typo: "start" instead of "sart"

Isn't there a waiting period of multiple months to cash out in order to avoid long-range attacks?

Reply:
That is my understanding too. This is how the logic was in PoS originally. I thought the quarantine period was for that purpose.